### PR TITLE
[external-module-controller] modules update policy

### LINKED ISSUE #6560: Release ModuleReleaseUpdate crd for source modules
### Preflight Checklist

- [x] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Use case. Why is this important?

Control which modules we have to set

### Proposed Solution

We need to realize a crd like this:
```yaml
apiVersion: deckhouse.io/v1alpha1
kind: ModuleUpdateSettings
spec:
  releaseChannel: xxxx
  update:
    windows: {}
    mode: Auto | Manual
  moduleReleaseSelector:
    labelSelector:
      source: deckhouse
```

From module source we have to create ModuleRelease only if have some MRU set for it.
Apply ModuleRelease based on MRU settings

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by miklezzzz**
## Description
Adds `ModuleUpdatePolicy` custom resource definition that is used to define `Module Releases` update policies (manual or auto, with update windows or without). Matching between a `ModuleRelease` and a `ModuleUpdatePolicy` happens by means of `ModuleReleaseSelectot` field of the `ModuleUpdatePolicy` (a label selector).  `Module Releases` not matching any `ModuleUpdatePolicy` can't be subject to deploying.  

`releaseChannel` field of `ModuleSource` is deprecated and release channels are defined by `releaseChannel` fields of `ModuleUpdatePolicyes`.

`ModuleSource` logic is also reworked in such a way that if there is no `Module Update Policy` matching an upcoming `Module Release` from a `Module Source` in the cluster, than that `Module Release` isn't going to be created from the `Module Source`, until a relevant `Module Update Policy` is created. If there are two or mode `Module Update Policies` matching an upcoming `Module Release` in the cluster than that `Module Release` is also not going to be created and en error will be reflected in the `Module Source` object saying that there is more than one policy, matching the release.

Also, external-module-manager `create_deckhouse_modules_source.go` hook has been updated to create default `Module Update Policy` for deckhouse `Module Source`, named deckhouse and matching `source=deckhouse` `Module Releases`.

## Why do we need it, and what problem does it solve?
Closes #6560

## What is the expected result?
An addition CR `ModuleUpdatePolicy` is available in a cluster and these entities are required for:
* creating module releases from module sources with release channels defined in module update policies;
* deploying module releases based on module update policies they match.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: external-module-controller
type: feature
summary: Add ModuleUpdatePolicy custom resource and implemet its logic.
impact: New modules releases aren't created and deployed without relevant module update policies in place. ModuleUpdatePolicy for the `deckhouse` ModuleSource is created automatically.
impact_level: high
```

### CODE REVIEW

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go [hunk 1]
```diff
@@ -80,8 +83,19 @@ type Controller struct {
 	restartReason string
 }
 
+const (
+	approvalAnnotation     = "release.deckhouse.io/approved"
+	defaultScanInterval    = 15 * time.Second
```
Comment (yalosev) [inline deckhouse-controller/pkg/controller/module-controllers/release/controller.go:88]: Let's rename this. For `defaultCheckInterval` or something like that

Comment (miklezzzz) [reply deckhouse-controller/pkg/controller/module-controllers/release/controller.go]: done

File: deckhouse-controller/crds/module-source.yaml [hunk 2]
```diff
@@ -35,6 +35,7 @@ spec:
                 releaseChannel:
                   type: string
                   description: Desirable default release channel for modules in the current source.
+                  x-doc-deprecated: true
                   default: "alpha"
```
[the remark this answers is not in the document]
Comment (miklezzzz) [reply deckhouse-controller/crds/module-source.yaml]: done

File: deckhouse-controller/crds/doc-ru-module-release.yaml [hunk 3]
```diff
@@ -35,6 +35,10 @@ spec:
           jsonPath: .status.phase
           type: string
           description: 'Показывает текущий статус релиза.'
+        - name: effective policy
```
Comment (yalosev) [inline deckhouse-controller/crds/doc-ru-module-release.yaml:38]:
```suggestion
        - name: update policy
```

Comment (miklezzzz) [reply deckhouse-controller/crds/doc-ru-module-release.yaml]: done

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/yandex-cloud'...

File: modules/005-external-module-manager/hooks/create_deckhouse_modules_source.go [hunk 4]
```diff
@@ -116,6 +116,32 @@ func createDeckhouseModuleSource(input *go_hook.HookInput) error {
...
+		},
+		Spec: v1alpha1.ModuleUpdatePolicySpec{
+			ModuleReleaseSelector: v1alpha1.ModuleUpdatePolicySpecReleaseSelector{
+				LabelSelector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"source": "deckhouse",
+					},
+				},
+			},
+			ReleaseChannel: "Stable",
```
Comment (yalosev) [inline modules/005-external-module-manager/hooks/create_deckhouse_modules_source.go:136]: let's put `releaseChannel` variable here

File: deckhouse-controller/pkg/controller/module-controllers/source/controller.go [hunk 5]
```diff
@@ -424,8 +466,13 @@ func (c *Controller) createModuleRelease(ctx context.Context, ms *v1alpha1.Modul
...
 		},
 		ObjectMeta: metav1.ObjectMeta{
-			Name:   fmt.Sprintf("%s-%s", moduleName, result.ModuleVersion),
-			Labels: map[string]string{"module": moduleName, "source": ms.Name, "release-checksum": checksum},
+			Name: fmt.Sprintf("%s-%s", moduleName, result.ModuleVersion),
+			Labels: map[string]string{
+				"module":               moduleName,
+				"source":               ms.Name,
+				"release-checksum":     checksum,
+				"module-update-policy": policyName,
```
Comment (yalosev) [inline deckhouse-controller/pkg/controller/module-controllers/source/controller.go:474]: Maybe we can make this label as `update-policy`, to make it a bit shorter

File: modules/005-external-module-manager/hooks/create_deckhouse_modules_source.go (hunk 4 above)
Comment (miklezzzz) [reply modules/005-external-module-manager/hooks/create_deckhouse_modules_source.go]: done

File: deckhouse-controller/pkg/controller/module-controllers/source/controller.go (hunk 5 above)
Comment (miklezzzz) [reply deckhouse-controller/pkg/controller/module-controllers/source/controller.go:474]: done

Comment (miklezzzz) [reply deckhouse-controller/pkg/controller/module-controllers/source/controller.go]: was renamed to modules.deckhouse.io/updaet-policy :carpentry_saw:

File: deckhouse-controller/crds/module-update-policy.yaml [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+          description: |
+            Defines the update settings for a module's release.
+          required:
+            - spec
+          properties:
+            spec:
+              type: object
+              required:
+                - update
+                - moduleReleaseSelector
```
Comment (nabokihms) [inline deckhouse-controller/crds/module-update-policy.yaml:34]: Probably, the module release selector can be absent if an administrator wants to completely disable this update policy.

**Review by nabokihms — APPROVED** [review]

File: deckhouse-controller/crds/module-update-policy.yaml (hunk 6 above)
Comment (yalosev) [reply deckhouse-controller/crds/module-update-policy.yaml:34]: I think we will add something like `mode: Never` lately, to disable the policy

**Pull request merged**